Dafny 4 edits to the Tutorial, FAQs and Error catalog (#3587)
dafny-lang/dafny: docs/HowToFAQ/onepage.md

Although Dafny tries to make the syntax of these sorts of things as similar as possible between expressions and statements, there are some differences. 
The following Dafny Power User note may be helpful in understanding the syntax: [Dafny Power User: Statement vs. Expression Syntax](http://leino.science/papers/krml266.html).

# Is there any difference between a method without a modifies clause and a function with a reads this clause?  I know that the latter you can use in expressions, but otherwise, is there anything the former can do that the latter can’t, for example?

## Question

> Is there any difference between a method without a `modifies` clause and a function method with a `reads this` clause?  I know that the latter you can use in expressionse.  Is there anything the former can do that the latter can’t, for example?

Reviewer: ```suggestion
Is there any difference between a method without a `modifies` clause and a function method with a `reads this` clause?  I know that the latter you can use in expressions.  Is there anything the former can do that the latter can’t, for example?
```

```dafny
trait Test {
 var i: int
}
class A {
  ghost const allTests: set<Test>
  predicate testHelper() reads allTests {
    forall t: Test <- allTests :: t.i == 1
  }
}
```

That way, if you want to assume anything about the Test classes that you are modeling extern, 
you only need to specify an axiom that says that whatever Test you have was in allTests, 
which could have been a pool of Test objects created from the start anyway, and then you can use your axiom. 
# How do I say 'reads if x then this\`y else this\`z'? Dafny complains about the 'this'.

## Question: 

How do I say 'reads if x then this\`y else this\`z'? Dafny complains about the 'this'.

## Answer:

Here is some sample code that show a workaround.

```dafny
trait Test {
  var y: int
  var z: int
  function {:opaque} MyFunc(x: bool): int
    reads (if x then {this} else {})`y, (if x then {} else {this})`z {
    if x then y else z
  }
}

```
# How do I model extern methods that return objects?

## Question: 

How do I model extern methods that return objects?

## Answer:

When modeling extern functions that return objects, it's usually not good to have specifications that return objects. 

> It's better to have a predicate that takes the input of a function, an object, and relates the two together.

Reviewer: ```suggestion
It's better to have a predicate that takes the input of a function, an object, and relates the two with each other.
```

For example:

```dafny
trait {:extern} {:compile false} Test {
  var i: int
  var y: int
}
trait {:extern} {:compile false} Importer {
  function Import(i: int): (r: Test)
    ensures r.i == i

  method {:extern} {:compile false} DoImport(i: int) returns (r: Test)
    ensures r == Import(i)

  predicate Conditions(i: int) {
     && var r := Import(i);
     && r.y == 2
  }
}
```

> In this case, it's better to write a predicate, and use existential quantifiers along with the :| operator, 

Reviewer: ```suggestion
In this case, it's better to write a predicate, and use existential quantifiers along with the `:|` operator, 
```

> and there no need to prove uniqueness because we are in ghost code!

Reviewer: ```suggestion
and there is no need to prove uniqueness because we are in ghost code!
```

```dafny
trait {:extern} {:compile false} Test {
  var i: int
}
trait {:extern} {:compile false} Importer {
  predicate IsImported(i: int, r: Test) {
    r.i == i

>   }

Reviewer: ```suggestion
  }
  
```
